go/libraries/utils/svcs/controller.go

```diff
@@ -0,0 +1,189 @@
+//
+// Any attempt to register a service after |Start| or |Stop| has been called
+// will return an error.
+//
+// If an error occurs when initializing the services of a Controller, the
+// Stop functions of any already initialized Services are called in
+// reverse-order. The error which caused the initialization error is returned.
+//
+// In the case that all Services Init successfully, the error returned from
+// |Start| is the first non-nil error which is returned from the |Stop|
+// functions, in the order they are called.
+//
+// WaitForStart() can be called at any time on a Controller. It will
+// block until |Start| is called. After |Start| is called, if all the services
+// succesfully initialize, it will return |nil|. Otherwise it will return the
+// same error |Start| returned.
+//
+// WaitForStop() can be called at any time on a Controller. It will block
+// until |Start| is called and initialization fails, or until |Stop| is called.
+// It will return the same error which |Start| returned.
+type Controller struct {
+	mu        sync.Mutex
+	services  []*Service
+	initErr   error
+	stopErr   error
+	started   bool
+	startCh   chan struct{}
+	stopped   bool
+	stopCh    chan struct{}
+	stoppedCh chan struct{}
+}
+
+func NewController() *Controller {
+	return &Controller{
+		startCh:   make(chan struct{}),
+		stopCh:    make(chan struct{}),
+		stoppedCh: make(chan struct{}),
+	}
+}
+
+func (c *Controller) WaitForStart() error {
+	<-c.startCh
+	c.mu.Lock()
+	err := c.initErr
+	c.mu.Unlock()
+	return err
+}
+
+func (c *Controller) WaitForStop() error {
+	<-c.stoppedCh
+	c.mu.Lock()
+	var err error
+	if c.initErr != nil {
+		err = c.initErr
+	} else if c.stopErr != nil {
+		err = c.stopErr
+	}
+	c.mu.Unlock()
+	return err
+}
+
+func (c *Controller) Register(svc *Service) error {
+	c.mu.Lock()
+	if c.started {
+		c.mu.Unlock()
+		return errors.New("Controller: cannot Register a service on a controller which was already started")
+	}
+	c.services = append(c.services, svc)
+	c.mu.Unlock()
+	return nil
+}
+
+func (c *Controller) Stop() error {
+	c.mu.Lock()
+	if !c.started {
+		c.mu.Unlock()
+		return errors.New("Controller: cannot Stop a controller which was never started")
+	}
+	if c.stopped {
+		c.mu.Unlock()
+		return errors.New("Controller: cannot Stop a controller which was already stopped or which failed to initialize all its services")
+	}
+	c.stopped = true
+	close(c.stopCh)
+	c.mu.Unlock()
+	<-c.stoppedCh
+	return nil
+}
+
+func (c *Controller) Start(ctx context.Context) error {
+	c.mu.Lock()
+	if c.started {
+		return errors.New("Controller: cannot start service controller twice")
+	}
+	c.started = true
+	svcs := make([]*Service, len(c.services))
+	copy(svcs, c.services)
+	c.mu.Unlock()
+	for i, s := range svcs {
+		if s.Init == nil {
+			continue
+		}
+		err := s.Init(ctx)
+		if err != nil {
+			for j := i - 1; j >= 0; j-- {
+				if svcs[j].Stop != nil {
+					svcs[j].Stop()
+				}
+			}
+			c.mu.Lock()
```

Comment:
This seems to have a weird contract with Stop(). In particular, Stop() claims that its not safe to call it until after WaitForStart() returns successfully, but some calls to `Stop()` before `WaitForStart()` returns will succeed, even if `WaitForStart()` is going to return an error. However, if an Init function fails, it will write `stopped = true` here and then the call to `Stop` will fail.

Should `Stop` be changed to return an error in those cases as well? I.e., the Controller has to be fully inited in order for calling Stop() to not return an error?

Reply:
Changed this so that Stop never returns an error and can be called multiple times and concurrently.